File: voss/watchdog.py
"""Independent watchdog and kill path (Voss RFC 7.4, Binding 4.7).

The watchdog does not depend on the worker it supervises.  Kill order:
stop new work -> revoke capabilities -> terminate the OS process.  State
preservation is best-effort and never delays containment.  Health and
`accepts_work` are consulted by the broker before every protected use.
"""
from __future__ import annotations

import subprocess
import threading
import time
from typing import Any, Callable, Dict, Optional

StateInfo = Any


class HealthReport:
    def __init__(self, ok: bool, detail: str = ""):
        self.ok = ok
        self.detail = detail

    def __bool__(self) -> bool:
        return self.ok


class Watchdog:
    """Supervisor that can suspend/terminate a worker and revoke grants."""

    def __init__(self, worker_id: Optional[str] = None):
        self._worker_id = worker_id
        self._suspended: set = set()
        self._procs: Dict[str, subprocess.Popen] = {}
        self._lock = threading.RLock()
        self._ok = True
        self._detail = "healthy"
        self._health_source: Optional[Callable[[], StateInfo]] = None
        self._revoke_fn: Optional[Callable[[str], int]] = None
        self._audit_fn: Optional[Callable[[str, str], None]] = None

    def attach(
        self,
        *,
        health_source: Optional[Callable[[], StateInfo]] = None,
        revoke_fn: Optional[Callable[[str], int]] = None,
        audit_fn: Optional[Callable[[str, str], None]] = None,
    ) -> None:
        self._health_source = health_source
        self._revoke_fn = revoke_fn
        self._audit_fn = audit_fn

    # -------------------------------------------------------------- status

    def health(self) -> HealthReport:
        reason = ""
        if self._health_source is not None:
            source = self._health_source()
            if isinstance(source, HealthReport):
                if not source.ok:
                    reason = source.detail
        return HealthReport(self._ok and not reason, self._detail + ("; " + reason if reason else ""))

    def accepts_work(self, worker_id: str) -> bool:
        return self._ok and worker_id not in self._suspended

    # --------------------------------------------------------------- control

    def register_process(self, worker_id: str, proc: subprocess.Popen) -> None:
        with self._lock:
            self._procs[worker_id] = proc

    def suspend(self, worker_id: str, reason: str = "drift") -> None:
        """Stop new work and withhold outputs; reversible containment."""
        with self._lock:
            self._suspended.add(worker_id)
        if self._audit_fn:
            self._audit_fn(worker_id, f"suspend: {reason}")

    def resume(self, worker_id: str) -> None:
        with self._lock:
            self._suspended.discard(worker_id)

    def kill(
        self, worker_id: str, reason: str = "operator", terminate_process: bool = True
    ) -> Dict[str, Any]:
        """Stop new work -> revoke capabilities -> terminate the process."""
        with self._lock:
            self._suspended.add(worker_id)

            revoked = 0
            if self._revoke_fn and terminate_process:
                try:
                    revoked = int(self._revoke_fn(worker_id))
                except Exception:  # containment must not be blocked
                    revoked = -1

            terminated = False
            proc = self._procs.get(worker_id)
            if proc is not None and terminate_process:
                terminated = self._terminate(proc)

        if self._audit_fn:
            self._audit_fn(worker_id, f"kill: {reason}")
        return {
            "worker_id": worker_id,
            "reason": reason,
            "capabilities_revoked": revoked,
            "process_terminated": terminated,
        }

    def _terminate(self, proc: subprocess.Popen) -> bool:
        if proc.poll() is not None:
            return False
        try:
            proc.terminate()
            try:
                proc.wait(timeout=2)
            except subprocess.TimeoutExpired:
                proc.kill()
                proc.wait(timeout=2)
        except OSError:
            return False
        return proc.poll() is not None


def sleep_now(duration: float) -> None:
    time.sleep(duration)

File: rust/src/watchdog.rs
//! In-process watchdog and kill path.
//!
//! Kill order is: stop new work, revoke capabilities, then terminate the
//! worker process. The supervisor lives in the trusted process, so it is
//! not an independent host. Containment does not edit code, memory, the
//! model, or policy.

use std::collections::BTreeMap;
use std::process::Child;
use std::sync::{Arc, Mutex};

pub struct HealthReport {
    pub ok: bool,
    pub detail: String,
}

type HealthSource = Arc<dyn Fn() -> HealthReport + Send + Sync>;
type RevokeCallback = Arc<dyn Fn(&str) -> i64 + Send + Sync>;
type AuditCallback = Arc<dyn Fn(&str, &str) + Send + Sync>;

struct WatchdogState {
    suspended: BTreeMap<String, ()>,
    processes: BTreeMap<String, Child>,
    ok: bool,
    detail: String,
    health_source: Option<HealthSource>,
    revoke: Option<RevokeCallback>,
    audit: Option<AuditCallback>,
}

pub struct Watchdog {
    state: Mutex<WatchdogState>,
}

impl Watchdog {
    pub fn new() -> Self {
        Self {
            state: Mutex::new(WatchdogState {
                suspended: BTreeMap::new(),
                processes: BTreeMap::new(),
                ok: true,
                detail: "healthy".to_string(),
                health_source: None,
                revoke: None,
                audit: None,
            }),
        }
    }

    pub fn attach(
        &self,
        health_source: impl Fn() -> HealthReport + Send + Sync + 'static,
        revoke: impl Fn(&str) -> i64 + Send + Sync + 'static,
        audit: impl Fn(&str, &str) + Send + Sync + 'static,
    ) {
        let mut state = self.state.lock().expect("watchdog lock");
        state.health_source = Some(Arc::new(health_source));
        state.revoke = Some(Arc::new(revoke));
        state.audit = Some(Arc::new(audit));
    }

    pub fn health(&self) -> HealthReport {
        let (ok, detail, source) = {
            let state = self.state.lock().expect("watchdog lock");
            (state.ok, state.detail.clone(), state.health_source.clone())
        };
        let mut reason = String::new();
        if let Some(source) = source {
            let report = source();
            if !report.ok {
                reason = report.detail;
            }
        }
        HealthReport {
            ok: ok && reason.is_empty(),
            detail: if reason.is_empty() {
                detail
            } else {
                format!("{detail}; {reason}")
            },
        }
    }

    pub fn accepts_work(&self, worker_id: &str) -> bool {
        let state = self.state.lock().expect("watchdog lock");
        state.ok && !state.suspended.contains_key(worker_id)
    }

    pub fn register_process(&self, worker_id: &str, process: Child) {
        let mut state = self.state.lock().expect("watchdog lock");
        state.processes.insert(worker_id.to_string(), process);
    }

    pub fn suspend(&self, worker_id: &str, reason: &str) {
        {
            let mut state = self.state.lock().expect("watchdog lock");
            state.suspended.insert(worker_id.to_string(), ());
        }
        self.audit(worker_id, &format!("suspend: {reason}"));
    }

    pub fn kill(&self, worker_id: &str, reason: &str, terminate_process: bool) -> KillReport {
        let revoke = {
            let mut state = self.state.lock().expect("watchdog lock");
            state.suspended.insert(worker_id.to_string(), ());
            if terminate_process {
                state.revoke.clone()
            } else {
                None
            }
        };
        let revoked = revoke.as_ref().map(|revoke| revoke(worker_id)).unwrap_or(0);
        let mut process = {
            let mut state = self.state.lock().expect("watchdog lock");
            if terminate_process {
                state.processes.remove(worker_id)
            } else {
                None
            }
        };
        let terminated = process.as_mut().is_some_and(terminate_child);
        self.audit(worker_id, &format!("kill: {reason}"));
        KillReport {
            worker_id: worker_id.to_string(),
            reason: reason.to_string(),
            capabilities_revoked: revoked,
            process_terminated: terminated,
        }
    }

    fn audit(&self, worker_id: &str, detail: &str) {
        let callback = {
            let state = self.state.lock().expect("watchdog lock");
            state.audit.clone()
        };
        if let Some(callback) = callback {
            callback(worker_id, detail);
        }
    }
}

impl Default for Watchdog {
    fn default() -> Self {
        Self::new()
    }
}

pub struct KillReport {
    pub worker_id: String,
    pub reason: String,
    pub capabilities_revoked: i64,
    pub process_terminated: bool,
}

fn terminate_child(process: &mut Child) -> bool {
    if process.try_wait().ok().flatten().is_some() {
        return false;
    }
    let _ = process.kill();
    process.wait().is_ok()
}
